_episodes/scatter.md

```diff
@@ -1,18 +1,141 @@
 ---
-title: "Scattering"
-teaching: 0
+title: "Scatter/Gather in a CWL Workflow"
+teaching: 5
 exercises: 0
 questions:
-- "Key question (FIXME)"
+- "How can I parallelize running of tools in a workflow?"
 objectives:
-- "explain what is meant by the _scatter_ pattern in workflow design, and how it differs from the similar concept of parallel execution"
-- "identify when the scatter pattern appears in a workflow description"
-- "(FIXME: does the above cover the intended meaning of the following two points from the lesson development sprint?); running the same program on each file; running the same program the same way except for one parameter"
+- "Learn how to scatter steps of a CWL workflow."
 keypoints:
-- "First key point. Brief Answer to questions. (FIXME)"
+- "Use"
 ---
-By the end of this episode,
-learners should be able to
-__implement scattering of steps in a workflow__.
+This workflow uses a real life example of open-source CWL CommandLineTools from Dockstore.
 
-{% include links.md %}
+CWL Workflows can scatter array inputs to run in parallel within the same step and gather all the outputs together to be used in subsequent steps. It's similar to GNU parallel in that it will execute all jobs in parallel using a single script and multiple input arguments.
+
+Use the ScatterFeatureRequirement in the requirements section of a workflow to begin enabling scattering in a single step.
+
+~~~
+requirements:
+   ScatterFeatureRequirement: {}
+~~~
+
+In the steps section, use scatter and the input variable to scatter on, in this example its the intervals, which uses the chromosomes string array specified in the inputs section of the Workflow.
+
+Exercise 1:
+
+Given this workflow, how would you parallelize GATK_HaplotypeCaller to run all chromosomes simultaneously?
+
+~~~
+cwlVersion: v1.0
+class: Workflow
+requirements:
+   ScatterFeatureRequirement: {}
+inputs:
+   bam: File
+   chromosomes: string[]
+outputs:
+  HaplotypeCaller_VCFs:
+    type: File[]
+    outputSource: GATK_HaplotypeCaller/vcf
+steps:
+  GATK_HaplotypeCaller:
+    run: GATK_HaplotypeCaller.cwl
+    in:
+      input_bam: bam
+      intervals: chromosomes
+    out: [vcf]
+~~~
```

Comment:
So currently this doesn't work, perhaps we should give them a working example with a single chromosome.